contracts/interchain-token/src/contract.rs

```diff
@@ -1,30 +1,220 @@
+            .get(&DataKey::TokenId)
+            .expect("token id not found")
     }
 
-    pub fn owner(env: &Env) -> Address {
+    pub fn interchain_token_service(env: &Env) -> Address {
         env.storage()
             .instance()
-            .get(&DataKey::Owner)
-            .expect("owner not found")
+            .get(&DataKey::InterchainTokenService)
+            .expect("interchain token service not found")
+    }
+
+    pub fn mint(env: Env, minter: Address, to: Address, amount: i128) -> Result<(), ContractError> {
+        minter.require_auth_for_args((&to, amount).into_val(&env));
+
+        check_nonnegative_amount(amount);
+
+        let is_authorized = env
+            .storage()
+            .persistent()
+            .get::<_, bool>(&DataKey::Minter(minter.clone()))
+            .unwrap_or(false);
+
+        if !is_authorized {
+            return Err(ContractError::NotAuthorizedMinter);
+        }
+
+        extend_instance_ttl(&env);
+
+        receive_balance(&env, to.clone(), amount);
+
+        TokenUtils::new(&env).events().mint(minter, to, amount);
+
+        Ok(())
     }
 
-    pub fn transfer_ownership(env: Env, new_owner: Address) {
-        let owner: Address = Self::owner(&env);
-        owner.require_auth();
+    pub fn set_admin(env: Env, new_admin: Address) {
+        let admin = admin(&env);
+        admin.require_auth();
+
+        extend_instance_ttl(&env);
 
-        env.storage().instance().set(&DataKey::Owner, &new_owner);
+        env.storage().instance().set(&DataKey::Admin, &admin);
+
+        TokenUtils::new(&env)
+            .events()
+            .set_admin(admin.clone(), new_admin.clone());
+
+        event::set_admin(&env, admin, new_admin);
+    }
+
+    pub fn add_minter(env: &Env, minter: Address) {
+        admin(env).require_auth();
+
+        env.storage()
+            .persistent()
+            .set(&DataKey::Minter(minter.clone()), &true);
+
+        event::add_minter(env, minter);
+    }
+}
+
+#[contractimpl]
+impl token::Interface for InterchainToken {
+    fn allowance(env: Env, from: Address, spender: Address) -> i128 {
+        extend_instance_ttl(&env);
+        read_allowance(&env, from, spender).amount
+    }
+
+    fn approve(env: Env, from: Address, spender: Address, amount: i128, expiration_ledger: u32) {
+        from.require_auth();
+
+        check_nonnegative_amount(amount);
+        extend_instance_ttl(&env);
+
+        write_allowance(
+            &env,
+            from.clone(),
+            spender.clone(),
+            amount,
+            expiration_ledger,
+        );
+
+        TokenUtils::new(&env)
+            .events()
+            .approve(from, spender, amount, expiration_ledger);
+    }
+
+    fn balance(env: Env, id: Address) -> i128 {
+        extend_instance_ttl(&env);
+        read_balance(&env, id)
+    }
+
+    fn transfer(env: Env, from: Address, to: Address, amount: i128) {
+        from.require_auth();
+
+        check_nonnegative_amount(amount);
```

Comment:
if this works with existing stellar tokens, then we don't need to check

Reply:
Even though the [Stellar native token](https://github.com/stellar/stellar-protocol/blob/76d203f3e73f2f78b95f300f32ccb813b71be556/core/cap-0046-06.md?plain=1#L475) supports handling negative amounts, I reviewed the Soroban examples and [other implementations](https://github.com/search?q=%22impl+token%3A%3AInterface+for%22+language%3ARust&type=code), and it appears they do not allow negative amounts in custom token implementations. Since interchain tokens are also custom tokens, I believe ensuring non-negative amounts is our decision. I would prefer to enforce this check to prevent potential security issues. Let me know your thoughts on this.

Reply:
we can revert for now, and change later if their team suggests